src/python/pants/backend/nfpm/util_rules/generate_config.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import yaml
+
+from pants.backend.nfpm.config import NfpmContent, file_info
+from pants.backend.nfpm.field_sets import NfpmPackageFieldSet
+from pants.backend.nfpm.fields.all import NfpmDependencies
+from pants.backend.nfpm.fields.contents import (
+    NfpmContentDirDstField,
+    NfpmContentDstField,
+    NfpmContentFileSourceField,
+    NfpmContentSrcField,
+    NfpmContentSymlinkDstField,
+    NfpmContentSymlinkSrcField,
+    NfpmContentTypeField,
+)
+from pants.backend.nfpm.target_types import NfpmContentFile
+from pants.core.goals.package import TraverseIfNotPackageTarget
+from pants.engine.fs import CreateDigest, DigestEntries, FileContent, FileEntry
+from pants.engine.internals.native_engine import Digest
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.unions import UnionMembership
+from pants.util.logging import LogLevel
+from pants.util.strutil import softwrap
+
+
+@dataclass(frozen=True)
+class NfpmPackageConfigRequest:
+    field_set: NfpmPackageFieldSet
+    content_sandbox_digest: Digest  # NfpmContentSandbox.digest
+
+
+@dataclass(frozen=True)
+class NfpmPackageConfig:
+    digest: Digest  # digest contains nfpm.yaml
+
+
+class InvalidNfpmContentFileTargetsException(Exception):
+    pass
+
+
+class NfpmSrcMissingFromSandboxException(Exception):
+    pass
+
+
+@rule(level=LogLevel.DEBUG)
+async def generate_nfpm_yaml(
+    request: NfpmPackageConfigRequest, union_membership: UnionMembership
+) -> NfpmPackageConfig:
```

Comment:
RuleRunner based integration test
Don't just test with the tests for the `package_nfpm_package` rule. Maybe these tests can create  intentionally problematic sandbox contents to ensure certain errors happen.
And, this might be the place to test that the `nfpm.yaml` file is well formed.